[SPARK-17017][MLLIB][ML] add a chiSquare Selector based on False Positive Rate (FPR) test

mpjlu opened the pull request:
## What changes were proposed in this pull request?

Univariate feature selection works by selecting the best features based on univariate statistical tests. False Positive Rate (FPR) is a popular univariate statistical test for feature selection. We add a chiSquare Selector based on False Positive Rate (FPR) test in this PR, like it is implemented in scikit-learn. 
http://scikit-learn.org/stable/modules/feature_selection.html#univariate-feature-selection
## How was this patch tested?

Add Scala ut

Commit:
add a chiSquare Selector based on False Positive Rate (FPR) test

Commit:
Merge remote-tracking branch 'origin/master' into fprChiSquare

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -197,3 +197,28 @@ class ChiSqSelector @Since("1.3.0") (
 }
+
+/**
+ * Creates a ChiSquared feature selector by False Positive Rate (FPR) test.
```
Is there any link to document what this means? I had to double-check it means what I think it means and could only find indirect references. We might very briefly explain it as a bound on the likelihood that the feature only by chance appears to be predictive, given the data, and even give an indicative common value like 0.05.

mpjlu replied:
Hi @srowen , http://blog.datadive.net/selecting-good-features-part-i-univariate-selection/, this is a link to introduce p-value for feature selection.

srowen replied:
Great, that's worth putting in the comment.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -197,3 +197,28 @@ class ChiSqSelector @Since("1.3.0") (
+  @Since("2.1.0")
+  def fit(data: RDD[LabeledPoint]): ChiSqSelectorModel = {
+    val indices = Statistics.chiSqTest(data)
+      .zipWithIndex.filter { case (res, _) => res.pValue < alpha }
```
It's trivial, but, `<=`? given the semantics?
This is also not related to your change, but this seems to return the features sorted by index, and not by importance, in both cases. That seems a little funny -- @avulanov do I read correctly that you wrote this code, and was it your intent?

mpjlu replied:
Here we don't need the features sorted by importance, because all the features are filtered by pvalue < alpha. But the ChiSqSelectorModel needs the features sorted by index.

srowen replied:
Yeah, what I should really say is: why does the model want sorted indices? it can sort them if needed internally. It's consuming just a set of indices, conceptually.

This, sort of, leads to maybe a better proposal. Why not retain the p values in the model by not keeping only the indices? Then, the model itself can be configured to do different types of selection. This would avoid having to make a second (or third or fourth) type of chi-squared selector. In fact all of them are doing the same selection under the hood, so it doesn't really make sense to make new selectors.

What's differing here is how the model uses the chi-squared fit to do feature selection, so I wonder if we can just push the change there? it may make this easier to implement in `.ml` too.

mpjlu replied:
Yes, great proposal. We can do that.  But the chiSquareSelector should be changed a lot to implement your proposal, I think it is necessary. I will submit a new PR based on my understanding of your proposal, how do you think about it? Thanks.

srowen replied:
I don't think it changes much at all, just that the model now contains the chi-squared results per feature, not just indices. Should be a couple lines, and no API change. The model changes internally though, but still not necessarily any user-visible API change. It can still expose the same info, it just has more information internally.

mpjlu replied:
The only problem here is how to the reuse the result of  "Statistics.chiSqTest(data)", so why not add a var to save the result.  we don't need to change the model related code.  Just need to add two lines of code to check which selection method is used. No API needs to change.  We need to add API to set parameters about other selection methods, like select by FPR. How do you think about it. @srowen

srowen replied:
Yes that also makes sense and works. One down side is that if I for some reason wanted to do different types of chi-squared selection I'd have to make two (or later, more) models, when conceptually there's only one chi-squared test underlying all of them. That seemed a bit funny, but not crazy.

I suppose I'm thinking that a next logical step is to simply also expose the p-values to the caller as well. Right now there's no way to tell just how important the features are; you only know which were selected. And then, it wouldn't make sense to have different selectors because they'd all be doing the same work, so the logic would belong in the model. Is that compelling?

mpjlu replied:
Two or more models based on the same chi-squared test is reasonable, because the chi-squared test results contain "pValue,  degreesOfFreedom,  statistic,,,". numTopFeatures selection method uses statistic, and FPR selection method uses pValue. 

Expose the p-value to the caller maybe useful, but we can expose ChiSqTestResult, not only p-value.

srowen replied:
The model can even choose to hide `ChiSqTestResult` and expose a different API. That can be done later too. It just occurred to me that it might be useful to build towards that by letting the model class own this info and what is done with it.

mpjlu replied:
Do you mean we can put some selection logic or more selection results not only indices to the model (ChiSqSelectorModel) ?

srowen replied:
Yes I was thinking the model itself would get `ChiSqTestResult` and then could be configured to perform different types of selection in `transform` by a parameter to the model instance.

avulanov replied:
@srowen I've checked our thread with @mengxr on that feature https://github.com/apache/spark/pull/1484. 
  - We preserve the order of indexes to make the selection of features with one loop (i.e. linear time complexity). Here is the code: https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala#L74. The logic of feature selector, which is selection of N top features, does not imply that it will sort the features by their Chi-square value. A parameter must be introduced if it is required for some use-case.
  - We were planning to include Chi-square values in the model later if needed https://github.com/apache/spark/pull/1484#discussion_r23876952

@mpjlu It seems that FPR feature selection should not modify the code of existing `ChiSqSelector`, because FPR feature selection works on top of a scoring function rather than on top of another selector. Scoring function is a parameter, and it might be Chi-square. For example, please refer to Sklearn's `FPR` implementation mentioned. It uses ANOVA as a default scoring function http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFpr.html#sklearn.feature_selection.SelectFpr.

mpjlu replied:
Hi @srowen ,  if we configure the model instance with different parameter  to perform different types of selection, will that be inconsistent with the MLlib Estimator/Model style, and cause the user confused?  
If it is not a problem, I will submit a new PR to configure the model to perform different selection.

srowen replied:
Hm, would it be more consistent to add the parameter controlling what type of feature selection is done to `ChiSqSelector`? it would actually compute the same thing in all cases, and pass along the p-values as well as what type of selection it is supposed to do, to the `ChiSqModelSelector` instance, which would then behave accordingly.

This might be more consistent with other models and at the same time flexible. Then the model itself doesn't have a configurable parameter, but still contains all the raw information required to answer additional questions, like "what are the p-values for the features?"

mpjlu replied:
Thanks, this makes sense.  It is better to add the parameter controlling what type of feature selection is done to ChiSqSelector.  But my idea is the the selection is done in "fit" function to be consistent with other feature selection methods, we don't need to pass what type of selection to the model. If we consider the user can get the p-value, maybe return the p-value from ChiSqSelector is ok?    If so , we don't need to make any change to the model class. How do you think about it?

srowen replied:
Yeah, I suppose I _really_ thought it would be nice for the model to expose the p values to the caller as a side benefit, since they're the essence of the result. Exposing this isn't strictly required here though so I would not say I feel strongly about it. But that was the motive.

OK, then we're back to mostly what you're already proposing. What about "fixing" the model so that it sorts the indices internally, while we're here, then? also not strictly required but might make this a little cleaner.

mpjlu replied:
ok,  I will pass the p values to the model and expose to the caller, and let the model to sort the indices internally.  I will add the parameter controlling what type of feature selection is done in "fit" function of ChiSqSelector. Is my understanding right?  Thanks.

srowen replied:
Yes I personally think that is a good solution. It sounds consistent, enables the change here, and exposes additional useful info to the caller.

mpjlu replied:
ok, thanks. I will submit a new PR based on our discussion.

srowen commented:
This would also need to modify the implementation in `.ml` to somehow accommodate the new params.

mpjlu commented:
Hi, @srowen , I can modify the implementation in .ml to accommodate the new params. Thanks.

avulanov commented:
@srowen I've checked our thread with @mengxr on that feature https://github.com/apache/spark/pull/1484. 
  - We preserve the order of indexes to make the selection of features with one loop (i.e. linear time complexity). Here is the code: https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala#L74. The logic of feature selector, which is selection of N top features, does not imply that it will sort the features by their Chi-square value. A parameter must be introduced if it is required for some use-case.
  - We were planning to include Chi-square values in the model later if needed https://github.com/apache/spark/pull/1484#discussion_r23876952

@mpjlu It seems that FPR feature selection should not modify the code of existing `ChiSqSelector`, because FPR feature selection works on top of a scoring function rather than on top of another selector. Scoring function is a parameter, and it might be Chi-square. For example, please refer to Sklearn's `FPR` implementation mentioned. It uses ANOVA as a default scoring function http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFpr.html#sklearn.feature_selection.SelectFpr.

srowen commented:
OK, makes sense @avulanov though I'm not sure why the model can't sort the indices if it requires this as an internal detail. No big deal. After this change it may not matter. Conceptually though, the output of chi-squared feature selection does certainly imply an ordering because it computes a p-value for each feature. It's useful info and there's a use for it here now.

avulanov commented:
Yes, it seems that index sort can be done inside the model. With regards to the sort by p-value, I have taken a brief look at chi-squared feature selection in sci-kit and Weka, and they don't seem to sort the output.

mpjlu commented:
Hi  @avulanov .  In general, FPR feature selection should not modify the code of existing ChiSqSelector, as we have implemented in this PR.  But if we need to reuse the ChiSqTestResult (Statistics.chiSqTest(data)),  it is better to modify the code of ChiSqSelector.  

In Scikit-learn, for each SelectKBest, SelectFpr, SelectPercentile and so on, create an object for it, as we implemented in this PR. The good point of this method is it is consistent across the LIB, all use the same Estimator/Model style.  The disadvantage is it cannot reuse the results of score function. @srowen

Commit:
Configure the ChiSqSelector to reuse ChiSqTestResult by numTopFeatures, Percentile, and Fpr selector

Commit:
Config the ChiSqSelector to reuse the ChiSqTestResult by KBest, Percentile and FPR selector

Commit:
Merge branch 'master' into fprChiSquare2

Commit:
add Since annotation

mpjlu commented:
Hi @srowen, I have added the parameter to control the feature selection type.
The usage is like this: 
**var selector = new ChiSqSelector()
var model = selector.fit(df) // by default, the selector is selection numTopFeatures (50)
var newModel = selector.selectKBest(10), or var newModel = selector.selectPercentile(5), or,,**
You can fit the DataFrame one time, and generate the model multi times. 

And the indices is sort in the model internally as we have discussed. 

For pass the p-value to the model function, this update does not include it. Because for the KBest and Percentile selection, the fit function uses ChiSqTestResult.statics to generate the model. For Fpr, the fit function uses ChiSqTestResult.p-value.  So it maybe better to pass ChiSqTestResult to the model and expose to the caller. And I think it is better to submit another PR for  "pass value to model and expose to the caller" problem. Because much codes will be changed for this problem, includes which data should be passed to the model, how  to save the model, how to test the model.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -67,9 +90,27 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
-  def setNumTopFeatures(value: Int): this.type = set(numTopFeatures, value)
+  @Since("2.1.0")
+  def setNumTopFeatures(value: Int): this.type = {
+    set(selectorType, "KBest")
```
Ideally, let's make these values like "KBest" some fields in a `private object`

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -54,6 +54,29 @@ private[feature] trait ChiSqSelectorParams extends Params
+  final val percentile = new DoubleParam(this, "percentile",
+    "Percentile of features that selector will select, ordered by statistics value descending.",
+    ParamValidators.gtEq(0))
+  setDefault(percentile -> 10)
```
I understand this means 10%, but I would expect it to be specified as 0.1. I wonder how it's done elsewhere? I think we've not used percents. At least we also want to validate that the value is <= 100% (or 1)

mpjlu replied:
I will update, thanks

srowen replied:
I think we still need to change this, because other functions like the SQL percentile function accept values in [0,1] instead of [0,100]

mpjlu replied:
ok, I will update it. Thanks.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -91,8 +132,38 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
       }
-    val chiSqSelector = new feature.ChiSqSelector($(numTopFeatures)).fit(input)
-    copyValues(new ChiSqSelectorModel(uid, chiSqSelector).setParent(this))
+    $(selectorType) match {
```
I may be missing something simple but I don't see why this `ChiSqSelector` makes a new `ChiSqSelector` to do the fitting. Doesn't it just use itself?

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -171,14 +177,47 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
+  @Since("1.3.0")
+  def this(numTopFeatures: Int) {
+    this()
+    this.numTopFeatures = numTopFeatures
```
This should call `setNumTopFeatures` to set the type too?

mpjlu replied:
This is not necessary, because the default selectorType is KBest

srowen replied:
OK. It seemed to split the logic a bit here but it's not bad. The default behavior needs to be documented then. Now there is effectively a default numTopFeatures.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +228,35 @@ class ChiSqSelector @Since("1.3.0") (
+      case ChiSqSelectorType.KBest => selectKBest(numTopFeatures)
+      case ChiSqSelectorType.Percentile => selectPercentile(percentile)
+      case ChiSqSelectorType.Fpr => selectFpr(alpha)
+      case _ => throw new Exception("Unknown ChiSqSelector Type")
```
`IllegalStateException` is probably better and this can include the type in the message.

mpjlu replied:
Thanks, I will update

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +228,35 @@ class ChiSqSelector @Since("1.3.0") (
+  }
+
+  @Since("2.1.0")
+  def selectKBest(value: Int): ChiSqSelectorModel = {
```
I would think these would be private? `fit` is the entry point. They can receive the results as an argument?

mpjlu replied:
The user can call selectKBest directly, because, use can fit one time, and generate models by selectKBest multi times.

srowen replied:
Oh I thought this was the discussion we had, that the model would still just expose `fit` and `transform` for consistency. It's not bad to expose these methods, just wondered if it were consistent. I thought that was the reasoning to not do it.

mpjlu replied:
Now, the user can use fit and transform only for consistent. 
If the user wants to reuse ChiSqTestResult, call selectKBest, SelectPercentile will help them. If we don't consider reuse ChiSqTestResult, we can not expose this methods.

srowen replied:
Yeah I thought the idea is that the model would be fixed, without parameters, though it would have all the information internally (the chi squared results) to select or re-select features to filter in based on different logic. And that the selector would expose the parameters. I'm not against your approach here just thought that's what you were saying would be inconsistent. I suppose you're right that the model would have to mutate its state to compute different sets of indices and that introduces a complication.

mpjlu replied:
The usage scenario here is:
model = selector.fit(df)
result = model.transform(features).
I try to understand your proposal:
Suppose the user generate the model1 use KBest:
model1 = selector.fit(df). 
you mean the user can config selector, like 
selector.setPercentile(0.1)
then, the model1 will be a different model, like it is generated by Percentile method, because it can use the parameters of selector.  Does my understanding right?

srowen replied:
Yeah I assume that the selector is configured to fit a certain type of model -- top k, percentile, etc -- and then produces that model, which is fixed and of a certain type. So the type is selected as a parameter to the selector, not the model. I think you were suggesting that is how other selectors work and I think you're right.

mpjlu replied:
Thanks. I am suggesting how one selector can generate multi models by only fit one time, because we want to reuse the ChiSqTestResult, this result is generated in fit(DataFrame). So we want to only fit one time and can generate multi models.   
According to your proposal, suppose we want to generate KBest with defaults numTopFeateatures. we can do like this:
val selector = new ChiSqSelector()          (1)
val model = selector.fit(dataframe)         (2)
then the user want to generate KBest model with only top 20 features.  The user can do like this:
selector.setNumToFeatures(20)     //here is the same selector as (1) , and the model (2) now is with Top 20 features, because it can use the parameters of selector. This is my understanding of your proposal, is my understanding right?   Thanks very much.

srowen replied:
The chi squared test result should be the same no matter which type of selection you do, yes. I agree it's desirable to avoid computing it many times if one wanted to perform many variations on chi squared selection, but is that a common use case?

Still, I can also imagine going back to the idea of only finding the chi-squared values in the selector, and letting the model instance be parameterized. So its behavior as a transformer would be modified by setting parameters on it. I'm OK with that, I just thought you were saying this wasn't consistent with other model classes. Maybe I misunderstood that.

mpjlu replied:
Hi @srowen , maybe I misunderstood you. I think make the model instance be parameterized is the same as current solution (make selector instance be parameterized). This only difference of your suggestion is moving the logic from selector to model.  Both of the solution is to avoid computing ChiSqTestResult many times when one wanted to perform different chi squared selection.

If we don't consider to avoid computing ChiSqTestResult many times,  the problem will be very easy, and can be consistent exactly with other methods.   Can we don't consider reuse ChiSqTestResult in this PR.  Thanks

srowen replied:
I agree with your second paragraph. I don't think it's very common that someone would perform different chi-squared selections on the same values but with different criteria. Therefore I'm not concerned about recomputing different models for different criteria. So I think indeed the model does not need parameters, and then it's consistent. Does that match your idea?

mpjlu replied:
Yes, I also think so.  I will update this PR. Thanks.

mpjlu replied:
Hi @srowen , I have updated this PR. Thanks.

srowen commented:
I think this will require a little update to the Python API to match. Not sure about SparkR

mpjlu commented:
Hi @srowen ,  I will update the Python API to match this changes. Now, the current Python API is not conflict with the changes.

Commit:
Not reuse the ChiSqTestResult to be consistent with other methods

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -54,6 +55,29 @@ private[feature] trait ChiSqSelectorParams extends Params
+
+  final val selectorType = new Param[String](this, "selectorType",
+    "ChiSqSelector Type: KBest, Percentile, Fpr")
+  setDefault(selectorType -> ChiSqSelectorType.KBest.toString)
```
Somewhere, there should be a few brief sentences describing how the types relate to the parameters to this class.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -91,8 +130,16 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
+        new feature.ChiSqSelector().setPercentile($(percentile)).fit(input)
+      case "Fpr" =>
+        new feature.ChiSqSelector().setAlpha($(alpha)).fit(input)
+      case _ => throw new IllegalStateException("Unknown ChiSqSelector Type.")
```
Put the invalid type itself into the error, and you can use string interpolation here.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -69,21 +73,22 @@ class ChiSqSelectorModel @Since("1.3.0") (
+   * @param filterIndices indices of features to filter
    */
   private def compress(features: Vector, filterIndices: Array[Int]): Vector = {
+    val orderedIndices = filterIndices.sorted
```
This can be computed once and stored, rather than store unsorted indices and resort them.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -32,17 +32,21 @@ import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.{Row, SparkSession}
 
+@Since("2.1.0")
+object ChiSqSelectorType extends Enumeration {
```
Can be `private[spark]`? I would expect this would be in `spark.ml` but I don't feel strong about it. It is referenced in both packages.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -91,8 +130,16 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
-    val chiSqSelector = new feature.ChiSqSelector($(numTopFeatures)).fit(input)
-    copyValues(new ChiSqSelectorModel(uid, chiSqSelector).setParent(this))
+    var model = $(selectorType) match {
+      case "KBest" =>
```
Shouldn't these use the toString of the new enumeration or something like that? or just parse the string as an enum value and match that.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +227,20 @@ class ChiSqSelector @Since("1.3.0") (
-      .sorted
+    var indices = selectorType match {
+      case ChiSqSelectorType.KBest => Statistics.chiSqTest(data)
+        .zipWithIndex.sortBy { case (res, _) => -res.statistic }
```
I think we want to compute and sort the chi squared stats in all cases, not just two. This can be lifted out.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +227,20 @@ class ChiSqSelector @Since("1.3.0") (
+        .map { case (_, indices) => indices }
+      case ChiSqSelectorType.Percentile => Statistics.chiSqTest(data)
+        .zipWithIndex.sortBy { case (res, _) => -res.statistic }
+        .take((data.count() * percentile / 100).toInt)
```
Not x% of the data, but x% of the features, right?

mpjlu replied:
Yes, I will update the code. thanks

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +227,20 @@ class ChiSqSelector @Since("1.3.0") (
+        .map { case (_, indices) => indices }
+      case ChiSqSelectorType.Fpr => Statistics.chiSqTest(data)
+        .zipWithIndex.filter{ case (res, _) => res.pValue < alpha }
+        .map { case (_, indices) => indices }
```
Finally this is common to all branches and can be lifted out

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +227,20 @@ class ChiSqSelector @Since("1.3.0") (
+      case ChiSqSelectorType.Fpr => Statistics.chiSqTest(data)
+        .zipWithIndex.filter{ case (res, _) => res.pValue < alpha }
+        .map { case (_, indices) => indices }
+      case _ => throw new IllegalStateException("Unknown ChiSqSelector Type")
```
Log type

mpjlu replied:
Hi @srowen , I have updated the code per all you comments.  Thanks very much.

Commit:
fix Percentile bugs, optimize the code

Commit:
change the default value of Percentile

hqzizania commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -171,14 +180,47 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
+
+  @Since("2.1.0")
+  def setPercentile(value: Double): this.type = {
+    require(value <= 1 && value >= 0, "Percentile should be larger than 0 and less than 100")
```
it should be less than 1.0 instead of 100 after, since you set its range [0, 1]

mpjlu replied:
Thanks, the code is updated

hqzizania commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -171,14 +180,47 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
+  }
+
+  @Since("2.1.0")
+  def setAlpha(value: Double): this.type = {
```
Does it need a `require` for alpha here?

mpjlu replied:
require is added, thanks

Commit:
Add require for setAlpha value

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -54,6 +55,36 @@ private[feature] trait ChiSqSelectorParams extends Params
+
+  final val percentile = new DoubleParam(this, "percentile",
+    "Percentile of features that selector will select, ordered by statistics value descending.",
+    ParamValidators.gtEq(0))
```
Use `.inRange(0, 1)` here? it needs to be <= 1 too.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -54,6 +55,36 @@ private[feature] trait ChiSqSelectorParams extends Params
+
+  final val alpha = new DoubleParam(this, "alpha",
+    "The highest p-value for features to be kept.",
+    ParamValidators.gtEq(0))
```
Same, only makes sense for this to be in [0,1]

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -91,8 +137,17 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
       }
-    val chiSqSelector = new feature.ChiSqSelector($(numTopFeatures)).fit(input)
-    copyValues(new ChiSqSelectorModel(uid, chiSqSelector).setParent(this))
+    var model = ChiSqSelectorType.withName($(selectorType)) match {
```
Although I don't feel strongly about it, we could avoid duplicating the "new feature.ChiSqSelector()" and "fit(input)" easily here.

mpjlu replied:
If avoid duplicating  the "new feature.ChiSqSelector()" and "fit(input)" here, the usage will be not consistent with other method, as we did before

srowen replied:
I just mean something like
```
val selector = new feature.ChiSqSelector()
ChiSqSelectorType.withName($(selectorType)) match {
  case ChiSqSelectorType.KBest => selector.setNumTopFeatures($(numTopFeatures))
  ...
}
val model = selector.fit(input)
...
```
Is that inconsistent with something?

mpjlu replied:
Sorry, I misunderstand your last comment.  Just think about reuse selector. 
This will not be inconsistent with other methods. I will update the code, thanks.

mpjlu replied:
Hi @srowen , I have updated the code. Thanks.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -69,21 +73,26 @@ class ChiSqSelectorModel @Since("1.3.0") (
+   * @param filterIndices indices of features to filter
    */
   private def compress(features: Vector, filterIndices: Array[Int]): Vector = {
+    val orderedIndices = if (isSorted(filterIndices)) {
```
I don't think there's any point in checking whether it's sorted. `.sorted` would ultimately do about the same thing and do no work if sorted already, so it doesn't add much. Plus you can remove the isSorted method too.

mpjlu replied:
ok, I agree with you. thanks.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +232,21 @@ class ChiSqSelector @Since("1.3.0") (
+    val chiSqTestResult = Statistics.chiSqTest(data)
+    val features = selectorType match {
+      case ChiSqSelectorType.KBest => chiSqTestResult
+        .zipWithIndex.sortBy { case (res, _) => -res.statistic }
```
It should be sorted in all cases I believe.

mpjlu replied:
Fpr doesn't need sort here.

srowen replied:
Yeah it's true, though in general the result makes sense to sort. Then again I guess the model right now doesn't make use of that ordering. It might later. I don't feel strongly about it but seems simpler to always sort.

mpjlu replied:
Hi @srowen , I have updated the code, do I miss any of your comments in this pr?

srowen replied:
I think it would still be better to sort here since it's actually simpler and not expensive. Do you feel strongly about it?

Otherwise, I have just one last thing I noted when reading all the code again -- how about changing "Fpr" to "FPR" because it does stand for "False Positive Rate"?

mpjlu replied:
Hi @srowen , do you mean we should move the sort from the model to here?
There are two sort here. 
One is sort the indices, which we have been moved from here to the model. 
The second is sort the features, because the KBest, and Percentile is selection by the order of  res.statistic, so we sort the features here.  FPR is selection by res.p-value > alpha, so don't need sort here.

mpjlu replied:
I will update Fpr to FPR, thanks very much.

srowen replied:
No just execute ` .zipWithIndex.sortBy { case (res, _) => -res.statistic }` no matter what before the match. I know it ends up being superfluous in one case but later we may want to retain sorted chi square results in the model. The other two cases need it anyway.

mpjlu replied:
Thanks, got you. Will update the code

mpjlu replied:
Hi @srowen , code is updated. Thanks.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -189,11 +232,21 @@ class ChiSqSelector @Since("1.3.0") (
+      case ChiSqSelectorType.Fpr => chiSqTestResult
+        .zipWithIndex.filter{ case (res, _) => res.pValue < alpha }
+      case errorType =>
+        throw new IllegalStateException("Unknown ChiSqSelector Type: $errorType")
```
needs "s" to use string interpolation

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -171,14 +180,48 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
+
+  @Since("2.1.0")
+  def setPercentile(value: Double): this.type = {
+    require(value <= 1 && value >= 0, "Percentile should be larger than 0 and less than 1")
```
Nit: suggest `require(0.0 <= value && value <= 1.0, "Percentile must be in [0,1]")` for compactness and consistency, and for alpha too below.

mpjlu replied:
thanks, will update

Commit:
rm isSorted function, change gtEq(0) to inRange(0,1) for percentile and alpha check

Commit:
Optimize fit function of ml ChiSqSelector

Commit:
Fpr to FPR, sort all cases in fit

srowen commented:
OK, that looks solid for Scala. I think we need to update the Python API too. That much should be straightforward. Are you in a position to do that -- I think we can call in advice from people that work on the Python API.

mpjlu commented:
Sure, I can update the Python API.

Commit:
Add Python API for ChiSqSelector

mpjlu commented:
Hi @srowen , I have added Python API and test cases for ChiSqSelector. Could you kindly review it again. Thanks.

srowen commented on python/pyspark/mllib/feature.py:
```diff
@@ -276,24 +276,64 @@ class ChiSqSelector(object):
-    def __init__(self, numTopFeatures):
-        self.numTopFeatures = int(numTopFeatures)
+    def __init__(self):
+        self.param = 50
```
It seems like param is used to mean many different things. Why not different fields like in the Scala version for clarity?

mpjlu replied:
Use three variables are clear. It needs another selectionType variable also.
In the fit function, according to the selection type, call different functions. If you prefer that, I will change the code. I am ok for both methods.  Thanks.

mpjlu replied:
Hi @srowen , how about adding some comments about param for clarity. Because it is very simple  code, only some set functions. Should be not very confusing to use param to mean different things.

srowen replied:
Hm, at least comments yes. But is it a problem to just use different fields for different values? they don't even have the same type.

mpjlu replied:
Hi @srowen , use different fields for different value is not a problem, just need another selectionType field, and in the fit function, the code will be:
if(selectionType == KBest)
   callMLlibFunc("fitChiSqSelectorKBest", self.numTopFeatures, data)
elseif(selectionType == Percentile)
   callMLlibFunc("fitChiSqSelectorKPercentile", self.percentile, data)
elseif(selecitonType == FPR)
   callMLlibFunc("fitChiSqSelectorFPR", self.alpha, data)
Is that ok? Thanks.

srowen replied:
I think that's clearer, yeah.

mpjlu replied:
Hi @srowen , I have updated the code. Could you please review it again? Thanks.

Commit:
split the ChiSqSelector param to numTopFeateres, Percentile, Alpha in Python

srowen commented on python/pyspark/mllib/feature.py:
```diff
@@ -271,29 +271,74 @@ def transform(self, vector):
+    """
+    This class defines the selector types of Chi Square Selector.
+    """
+    KBest, Percentile, FPR = range(3)
```
I see, this is the only way to express enums in Python 2?

mpjlu replied:
Hi @srowen , there are many ways to express enums in Python 2, this is the most frequently used and simplest way.

srowen commented on python/pyspark/mllib/feature.py:
```diff
@@ -305,7 +350,12 @@ def fit(self, data):
+            jmodel = callMLlibFunc("fitChiSqSelectorKBest", self.numTopFeatures, data)
+        elif self.selectorType == ChiSqSelectorType.Percentile:
+            jmodel = callMLlibFunc("fitChiSqSelectorPercentile", self.percentile, data)
+        elif self.selectorType == ChiSqSelectorType.FPR:
```
I'm not sure what the convention in on the Pyspark side for args checking. For example here it doesn't check that the type is one of the supported types. I guess follow the same convention you find elsewhere.

mpjlu replied:
Thanks  @srowen , the type checking is added.

srowen commented:
I think this is OK as far as I know though I'd appreciate it if any Python people can look at the Python side.

Commit:
Add type check for Python ChiSqSelector

Commit:
Change the exception type of value check

mpjlu commented:
Hi @yanboliang , could you please kindly review the python code of this PR. Thanks.

srowen commented:
@mpjlu I'm still following this one, and I looked again today. I couldn't see anything else to change. For a minute I thought we need to update ChiSqSelectorModelReader / ChiSqSelectorModelWriter code to also save new info, but, the model itself actually doesn't change. Just the process that selects the indices. I will commit tomorrow if there's no other comment.

mpjlu commented:
hi @srowen , thanks.  
This is my first PR. learn much from you. thanks very much.

SparkQA commented:
**[Test build #3257 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3257/consoleFull)** for PR 14597 at commit [`bbccac7`](https://github.com/apache/spark/commit/bbccac7b5d3812d71498f9a420829f3e67955794).

SparkQA commented:
**[Test build #3257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3257/consoleFull)** for PR 14597 at commit [`bbccac7`](https://github.com/apache/spark/commit/bbccac7b5d3812d71498f9a420829f3e67955794).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
change python code style

mpjlu commented:
Hi @srowen , Python style fail is updated. Thanks.

srowen commented:
Jenkins add to whitelist

SparkQA commented:
**[Test build #65314 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65314/consoleFull)** for PR 14597 at commit [`c35bcf1`](https://github.com/apache/spark/commit/c35bcf14a7c1694ec03f4c58fc6f5b8d5ef3e5bb).

SparkQA commented:
**[Test build #65314 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65314/consoleFull)** for PR 14597 at commit [`c35bcf1`](https://github.com/apache/spark/commit/c35bcf14a7c1694ec03f4c58fc6f5b8d5ef3e5bb).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65314/
Test FAILed.

SparkQA commented:
**[Test build #3259 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3259/consoleFull)** for PR 14597 at commit [`c35bcf1`](https://github.com/apache/spark/commit/c35bcf14a7c1694ec03f4c58fc6f5b8d5ef3e5bb).

SparkQA commented:
**[Test build #3259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3259/consoleFull)** for PR 14597 at commit [`c35bcf1`](https://github.com/apache/spark/commit/c35bcf14a7c1694ec03f4c58fc6f5b8d5ef3e5bb).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
change python code style

SparkQA commented:
**[Test build #65315 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65315/consoleFull)** for PR 14597 at commit [`e8f03ed`](https://github.com/apache/spark/commit/e8f03edbc54d3c9cb32688f79dd129dbf043da38).

SparkQA commented:
**[Test build #65315 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65315/consoleFull)** for PR 14597 at commit [`e8f03ed`](https://github.com/apache/spark/commit/e8f03edbc54d3c9cb32688f79dd129dbf043da38).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65315/
Test FAILed.

mpjlu commented:
hi @srowen , we have moved isSorted in ChiSqSelectorModel.  There is error message "*** method isSorted(Array[Int])Boolean in class org.apache.spark.mllib.feature.ChiSqSelectorModel does not have a correspondent in current version**". Should we revert it?
And we also need to add a new method numTopFeatures() in ChiSqSelector for the following error message: " *** method numTopFeatures()Int in class org.apache.spark.mllib.feature.ChiSqSelector does not have a correspondent in current version**" 
How do you think about it? Thanks.

Commit:
revert isSort to pass MiMa test

SparkQA commented:
**[Test build #65318 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65318/consoleFull)** for PR 14597 at commit [`ec74dde`](https://github.com/apache/spark/commit/ec74ddebe6bc9f1e85c60e9fcd88355691b5f42c).

SparkQA commented:
**[Test build #65318 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65318/consoleFull)** for PR 14597 at commit [`ec74dde`](https://github.com/apache/spark/commit/ec74ddebe6bc9f1e85c60e9fcd88355691b5f42c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65318/
Test PASSed.

mpjlu commented:
hi @srowen , I have updated the code for some MiMa test error. Could you please review it again. Thanks.

Commit:
Change MimaExcludes

SparkQA commented:
**[Test build #65357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65357/consoleFull)** for PR 14597 at commit [`6398f4c`](https://github.com/apache/spark/commit/6398f4ca954ff0971c5ef3db7956847b09d54849).

Commit:
Change Mima conflict

SparkQA commented:
**[Test build #65359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65359/consoleFull)** for PR 14597 at commit [`6cc4c92`](https://github.com/apache/spark/commit/6cc4c928c396d5e1aa0201814a2996d8ee1760e3).

SparkQA commented:
**[Test build #65357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65357/consoleFull)** for PR 14597 at commit [`6398f4c`](https://github.com/apache/spark/commit/6398f4ca954ff0971c5ef3db7956847b09d54849).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65357/
Test PASSed.

SparkQA commented:
**[Test build #65359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65359/consoleFull)** for PR 14597 at commit [`6cc4c92`](https://github.com/apache/spark/commit/6cc4c928c396d5e1aa0201814a2996d8ee1760e3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65359/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -68,8 +99,23 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
   /** @group setParam */
-  @Since("1.6.0")
-  def setNumTopFeatures(value: Int): this.type = set(numTopFeatures, value)
+  @Since("2.1.0")
```
Oh, I just noticed this: it should still say "since 1.6.0"

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -171,14 +166,48 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
-@Since("1.3.0")
-class ChiSqSelector @Since("1.3.0") (
-  @Since("1.3.0") val numTopFeatures: Int) extends Serializable {
+@Since("2.1.0")
```
Likewise the class is still "since 1.3.0"

srowen commented on python/pyspark/mllib/feature.py:
```diff
@@ -271,28 +271,75 @@ def transform(self, vector):
     .. versionadded:: 1.4.0
     """
-    def __init__(self, numTopFeatures):
+    def __init__(self):
```
Hm, does this actually remove the existing constructor or am I missing it? it should be possible to use the existing constructor that sets numTopFeatures, still.

Also I'm not sure what the relationship between versionadded and `@since` is, but seems like you're following the convention here.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/ChiSqSelector.scala:
```diff
@@ -171,14 +166,48 @@ object ChiSqSelectorModel extends Loader[ChiSqSelectorModel] {
+  var alpha: Double = 0.05
+  var selectorType = ChiSqSelectorType.KBest
+
+  @Since("1.3.0")
```
The existing constructor should still have javadoc maybe pointing to the setNumTopFeatures method to say that's the effect it has

srowen reviewed (changes requested):
The last thing to finish this off is to add a bit of documentation -- a brief summary of the three possible usages in the class doc. I don't think we need additional examples. That would at least help make sure people see the functionality.

mpjlu commented:
Thanks very much, I am in holiday now, will update the code this Sunday.

Commit:
add javadoc

SparkQA commented:
**[Test build #65559 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65559/consoleFull)** for PR 14597 at commit [`1d2f67f`](https://github.com/apache/spark/commit/1d2f67f01be415203173bc441fbb238ccb7121a9).

Commit:
fix mima conflict

SparkQA commented:
**[Test build #65563 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65563/consoleFull)** for PR 14597 at commit [`6220dd5`](https://github.com/apache/spark/commit/6220dd56378619bbee88432b94a378578bce0282).

SparkQA commented:
**[Test build #65559 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65559/consoleFull)** for PR 14597 at commit [`1d2f67f`](https://github.com/apache/spark/commit/1d2f67f01be415203173bc441fbb238ccb7121a9).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65559/
Test FAILed.

SparkQA commented:
**[Test build #65563 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65563/consoleFull)** for PR 14597 at commit [`6220dd5`](https://github.com/apache/spark/commit/6220dd56378619bbee88432b94a378578bce0282).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public static final class UnsignedPrefixComparatorNullsLast extends RadixSortSupport `
  * `  public static final class UnsignedPrefixComparatorDescNullsFirst extends RadixSortSupport `
  * `  public static final class SignedPrefixComparatorNullsLast extends RadixSortSupport `
  * `  public static final class SignedPrefixComparatorDescNullsFirst extends RadixSortSupport `
  * `abstract sealed class NullOrdering `
  * `case class SortOrder(`
  * `case class Length(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `trait Command extends LeafNode `
  * `trait RunnableCommand extends logical.Command `
  * `case class CreateTable(`
  * `case class AnalyzeCreateTable(sparkSession: SparkSession) extends Rule[LogicalPlan] `
  * `    class SetAccumulator[T] extends AccumulatorV2[T, java.util.Set[T]] `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65563/
Test PASSed.

srowen commented on python/pyspark/mllib/feature.py:
```diff
@@ -271,28 +271,75 @@ def transform(self, vector):
         return JavaVectorTransformer.transform(self, vector)
 
 
+class ChiSqSelectorType:
```
@MLnick do you have an opinion on the Python style?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala:
```diff
@@ -54,11 +55,44 @@ private[feature] trait ChiSqSelectorParams extends Params
 /**
  * Chi-Squared feature selection, which selects categorical features to use for predicting a
  * categorical label.
+ * The selector supports three selection methods: KBest, Percentile and FPR.
```
This is a good start but I think we could say some more. I suggest something like ...

The selector supports three selection methods: `KBest`, `Percentile`, and `FPR`. `KBest` chooses the _k_ top features according to a chi-squared test. `Percentile` is similar but chooses a fraction of all features instead of a fixed number. `FPR` chooses all features whose false positive rate meets some threshold.

Should this doc be applied to Python too?

mpjlu replied:
Thanks @srowen, I have updated the javadoc.

Commit:
Merge remote-tracking branch 'origin/master' into fprChiSquare

Commit:
change javadoc

SparkQA commented:
**[Test build #65585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65585/consoleFull)** for PR 14597 at commit [`88d2143`](https://github.com/apache/spark/commit/88d2143989a4219020100999c53bb1186fce5d1c).

SparkQA commented:
**[Test build #65585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65585/consoleFull)** for PR 14597 at commit [`88d2143`](https://github.com/apache/spark/commit/88d2143989a4219020100999c53bb1186fce5d1c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65585/
Test PASSed.

srowen commented:
This is ready to go ... except that it needs a rebase now, sorry.

Commit:
fix mima conflict

mpjlu commented:
No problem. thanks very much @srowen .

SparkQA commented:
**[Test build #65658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65658/consoleFull)** for PR 14597 at commit [`24f26f2`](https://github.com/apache/spark/commit/24f26f2393a5ba07d96b6a68b3fe33c7ae38778d).

SparkQA commented:
**[Test build #65658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65658/consoleFull)** for PR 14597 at commit [`24f26f2`](https://github.com/apache/spark/commit/24f26f2393a5ba07d96b6a68b3fe33c7ae38778d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65658/
Test PASSed.

srowen commented:
Nice one. I like this change. Merged to master.
